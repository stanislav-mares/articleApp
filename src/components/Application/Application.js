import React from 'react';
import PropTypes from 'prop-types';

//components
import Article from 'components/Article/Article';
import Comments from 'components/Comments/Comments';

export default class App extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      article: {},
      comments: []
    };
  }

  componentDidMount() {
    this.getArticle();
    this.getComments();
  }

  componentWillUnmount() {
    clearTimeout(this.timerArticle);
    clearTimeout(this.timerComments);
    clearTimeout(this.getMoreComments);
  }

  //simulation of get article request
  getArticle = (callback) => {
    this.timerArticle = setTimeout(
      () => {
        this.setState({article: document.__article});
        console.log('article downloaded');
        typeof callback === 'function' && callback(document.__article);
      },
      250
    );
  }

  //simulation of  get comments request
  getComments = (callback) => {
    this.timerComments = setTimeout(
      () => {
        this.setState({comments: document.__comments});
        console.log('comments downloaded');
        typeof callback === 'function' && callback(document.__comments);
      },
      250
    );
  }

  //simulation of get more comments request
  getMoreComments = (callback) => {
    this.timerMoreComments = setTimeout(
      () => {
        this.setState({
          comments: [ ...this.state.comments, ...document.__moreComments]
        });
        console.log('more comments downloaded');
        typeof callback === 'function' && callback([ ...document.__moreComments ]);
      },
      250
    );
  }

  render() {

    return(
      <main>
        <Article { ...this.state.article } />
        {Object.keys(this.state.article).length > 0 &&
          <Comments comments={this.state.comments} />}
        {this.state.comments.length > 0 &&
          <button onClick={this.getMoreComments} type="button">more comments</button>}
      </main>
    );
  }
}

App.propTypes = {
  /*props*/
};
