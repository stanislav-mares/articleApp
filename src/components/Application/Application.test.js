import 'jsdom-global/register'
import React from 'react';
import enzymeConfig from 'enzymeConfig/enzymeConfig';
import { mount, shallow } from 'enzyme';

import Application from 'components/Application/Application';

describe('<Application />', () => {

  document.__article = {
    author: 'Eliana Franco',
    date: '2017-09-10T22:00:05.919Z',
    text: `Your bones don't break, mine do. That's clear. Your cells react to bacteria and viruses differently than mine. You don't get sick, I do. That's also clear. But for some reason, you and I react the exact same way to water. We swallow it too fast, we choke. We get some in our lungs, we drown. However unreal it may seem, we are connected, you and I. We're on the same curve, just on opposite ends.`,
  };

  document.__comments = [
    { id: 1, author: 'Germán Toro', text: 'Great!', date: '2017-09-11T12:00:39.430Z' },
    { id: 2, author: 'Elena Builes', text: 'Awesome!', date: '2017-09-11T12:30:39.430Z' },
  ];

  document.__moreComments = [
    { id: 3, author: 'Germán Toro', text: 'Splendid!', date: '2017-09-11T14:30:05.919Z' },
    { id: 4, author: 'Elena Builes', text: 'Excelent!', date: '2017-09-11T11:30:05.919Z' },
  ];

  let componentShallow;
  let componentMount;

  beforeAll(() => {
    componentShallow = shallow(<Application />);
    componentMount = mount(<Application />);
  });

  it('Testing if no button (more comments) is rendered when no comment exists', () => {
    expect(componentMount.find('button').length).toBe(0);
  });

  it('Testing if getArtcile method works correctly', (done) => {

    const callback = (data) => {
      expect(componentMount.state().article).toEqual(data);
      done();
    }
    componentMount.instance().getArticle(callback);
  });

  it('Testing if getComments method works correctly', (done) => {

    const callback = (data) => {
      expect(componentMount.state().comments).toEqual(data);
      done();
    }
    componentMount.instance().getComments(callback);
  });

  it('Testing if getMoreComments method works correctly', (done) => {
    const callback = (newData) => {
      const expectedResult = [...prevBefore, ...newData];

      componentMount.state().comments.forEach((comment, i) => {
        const index = expectedResult.findIndex((item) => comment.id === item.id);
        expect(comment).toBe(expectedResult[index]);
      });
      done();
    }

    const prevBefore = componentMount.state().comments;
    componentMount.instance().getMoreComments(callback);
  });

  it('Testing structure of the App component', () => {
    expect(componentShallow).toMatchSnapshot();
  });
});
