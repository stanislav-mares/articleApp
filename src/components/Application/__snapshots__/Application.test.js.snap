// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Application /> Testing structure of the App component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Article />,
        false,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      false,
    ],
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Article />,
          false,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
      ],
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
